### DISCUSSION TITLE
Poor calibration result and depth map

### DISCUSSION THREAD

**Discussion by felipe8066** [decadenza/SimpleStereo#31] [Q&A]
Hello, I am having issues obtaining good results using the library. I have followed the examples, starting with the calibration process, for which I captured 40 photos of the chessboard in different positions. As an example, I'm attaching an image. When using the script "003 BuildStereoRig.py," I get the following results:

Left and rigth images:
![1_L](https://github.com/decadenza/SimpleStereo/assets/111525276/a0392598-67f3-487a-8d51-0b5278b7c5f7)

![1_R](https://github.com/decadenza/SimpleStereo/assets/111525276/5b612788-0dfe-492e-846b-22b06eca2776)

Reprojection error: 127.08500929243618
Centers: (array([0., 0., 0.]), array([ 442.39739036, -796.27468563, 851.04251781]))
Baseline: 1246.6122865943455

The error is high, but I'm currently unsure how to improve it. Moving forward, I apply the script for rig rectification, and finally, I run the script "011 Build3DPointCloud.py," resulting in a very poor outcome where nothing is visible. The issue seems to be with image rectification, but I'm unsure how to resolve it. I would appreciate your assistance as I've been stuck on this for a while and need to make progress with my work :(.

Disparity map:
![disparity](https://github.com/decadenza/SimpleStereo/assets/111525276/93a9788f-5a69-40b1-931c-fec5b03587e0)

Thanks in advance.

**Comment by decadenza** [comment]
Hi and thanks for writing here.

Could you please share the script you are running, so I can check the configuration parameters (like chessboard size)?

Are the images labelled in the correct order (the first image you posted seems the right one, the second seems the left one)?

Is the baseline large enough?

Cheers

**Comment by felipe8066** [comment]
Hello, thank you very much for responding. Below, I provide you with the code you requested.
```

import os
import simplestereo as ss
"""
Build a stereo rig object calculating parameters from calibration images.
"""

def extract_number(filename):
    return int(''.join(filter(str.isdigit, filename)))

left_images_path = os.path.join("/home", "felipe", "Documents", "Tesis", "Cuarta prueba lab", "Calibracion","Left")
right_images_path = os.path.join("/home", "felipe", "Documents", "Tesis", "Cuarta prueba lab","Calibracion", "Right")
save_file_path = os.path.join("/home", "felipe", "Documents", "Tesis", "Cuarta prueba lab", "rig.json")

left_images = [f for f in os.listdir(left_images_path) if f.endswith("_L.jpg")]
right_images = [f for f in os.listdir(right_images_path) if f.endswith("_R.jpg")]


if len(left_images) != len(right_images):
    raise ValueError("La cantidad de imágenes en las carpetas izquierda y derecha no coincide.")


left_images.sort(key=extract_number)
right_images.sort(key=extract_number)


images = [(os.path.join(left_images_path, left), os.path.join(right_images_path, right))
          for left, right in zip(left_images, right_images)]
print(f"Calibrating using {len(images)} images from:\nLeft images: {left_images_path}\nRight images: {right_images_path}...")


# Calibrate and build StereoRig object
rig = ss.calibration.chessboardStereo( images, chessboardSize=(7,7), squareSize=25)

# Save rig object to file
rig.save( save_file_path )

# Print some info
print("Saved in:", save_file_path)
print("Reprojection error:", rig.reprojectionError)
print("Centers:", rig.getCenters())
print("Baseline:", rig.getBaseline())

print("Done!")
```

I'll provide a diagram of how the cameras are arranged and why I determined that the left image is the one I shared above initially, and the right one is the second. Regarding the baseline inquiry, I'd like to mention that the cameras are separated by approximately 15 cm. I'm unsure if increasing the separation will improve the results. I would appreciate it if you could guide me on this to achieve better outcomes. Thank you.

![image](https://github.com/decadenza/SimpleStereo/assets/111525276/60247fa9-b957-4ef6-b39c-a01d8078013c)

**Reply by decadenza** [reply]
Hi and sorry for not noticing this before: you chessboard cannot be used for camera calibration, as **the number or rows must be different from the number of columns**.

You cannot have 7 rows == 7 columns, otherwise there is a 180 degrees uncertainty!

Also, I am pretty sure that the orientation of the images should not affect the stereo calibration and reconstruction, provided that is consistent. However, for the sake of simplicity, it might be worth rotating the images too.
Inverting left/right or up/down, instead, messes up with the disparity calculation.

**Comment by felipe8066** [comment]
Thank you for your response; I'll follow your instructions.

Based on how stereo vision works, I have some questions that could help me possibly improve the results:

- Can I rotate the chessboard for calibration images, or should I always keep the columns and rows in the same orientation?
- Regarding camera calibration, I am using two Nikon D3200 cameras with identical settings. Would it be sufficient to use the calibration script for both cameras, or should I perform calibration separately for each camera beforehand?
- About image rectification, I understand that the 3D reconstruction script incorporates image rectification, and it's not necessary to perform this process separately before running 3D reconstruction. Is this correct, or should I rectify the images before running the 3D reconstruction?

**Reply by decadenza** [reply]
>     Can I rotate the chessboard for calibration images, or should I always keep the columns and rows in the same orientation?

You can have any orientation, provided that rows != columns. It is actually better to move the chessboard in many different positions covering all the operating volume.

>     Regarding camera calibration, I am using two Nikon D3200 cameras with identical settings. Would it be sufficient to use the calibration script for both cameras, or should I perform calibration separately for each camera beforehand?

Calibrate stereo cameras directly. Stereo calibration will take care of both intrinsic parameters for each camera plus extrinsic ones (common, namely a rotation R and a translation t).

>     About image rectification, I understand that the 3D reconstruction script incorporates image rectification, and it's not necessary to perform this process separately before running 3D reconstruction. Is this correct, or should I rectify the images before running the 3D reconstruction?

You need to rectify before reconstruction. The way in which SS was designed is to have the following operations in sequence

1.  Calibrate your rig (represented as a `StereoRig` object) as shown in https://github.com/decadenza/SimpleStereo/blob/master/examples/003b%20BuildStereoRig.py
2. Convert your rig object to a `RectifiedStereoRig` object as shown in https://github.com/decadenza/SimpleStereo/blob/master/examples/005%20RectifyRig.py

At this point you can save your rig to file and load it when needed so you don't have to repeat the above operations every time.

3. Build a disparity map (https://github.com/decadenza/SimpleStereo/blob/master/examples/008%20StereoMatchingSGBM.py) from a couple of input images. Note that many stereo matching algorithm exist, and SS only implements a few basic ones.

4.  Convert the disparity information to a 3D point cloud (https://github.com/decadenza/SimpleStereo/blob/master/examples/011%20Build3DPointCloud.py), if needed.

I hope this helps. Good luck with you project, please share the results if you can!

**Comment by felipe8066** [comment]
Decadenza, thank you for your comments. I performed a new calibration following your instructions, which yielded better results. However, these results are not entirely satisfactory. I'm sharing the rectified images, which appear to have a somewhat spherical rectification, leading to poor outcomes. Additionally, I'm providing my set of calibration images in case you notice anything when reviewing them.

![LEFT Rectified_screenshot_18 01 2024](https://github.com/decadenza/SimpleStereo/assets/111525276/112d686f-ab96-4cc9-938e-4273ac34a9dd)
![RIGHT Rectified_screenshot_18 01 2024](https://github.com/decadenza/SimpleStereo/assets/111525276/e73b001a-2e68-472f-917b-be2b262a72e4)

Images for calibration:
https://drive.google.com/drive/folders/1iorj-JLsD6_64hzeHxKWuVtIxHnO7Ls2?usp=sharing

I would greatly appreciate your feedback to help achieve good results. Thank you.
Greetings.

**Reply by decadenza** [reply]
Please share your stereo rig object in JSON format too. So I can give a look to what happened exactly. There must be some problem with distortion coefficients...

**Comment by felipe8066** [comment]
The requested files are provided below:
[rig.json](https://github.com/decadenza/SimpleStereo/files/13990835/rig.json)
[rigRect.json](https://github.com/decadenza/SimpleStereo/files/13990836/rigRect.json)

**Comment by decadenza** [comment]
Hi @felipe8066 

I had a look at your files and there is no major problem. The effect you are experiencing is due to low number of calibration images, that do not allow the distortion coefficient to converge properly.

In fact, as a quick and dirty fix, you can see that in the following script I ignored the distortion coefficients and the result is definitely acceptable.

**I suggest to use >100 quality images to calibrate. The more the better (if the quality is good). Be sure to cover all the working volume.**

```
import sys
import os

import numpy as np
import cv2

import simplestereo as ss

# Paths
curPath = os.path.dirname(os.path.realpath(__file__))
loadPath = os.path.join(curPath,"img")    # Image folder
loadFile = os.path.join(curPath,"rigRect.json")

# Save rig object to file
rigRect = ss.RectifiedStereoRig.fromFile(loadFile)

# Read right and left image (please ensure the order!).
# Using the first two images used for calibration as demo.
img1 = cv2.imread(os.path.join(loadPath, "1_L.jpg"))
img2 = cv2.imread(os.path.join(loadPath, "1_R.jpg"))

###################################
# IMPORTANT Quick and dirty fix!!!!
rigRect.distCoeffs1 = None
rigRect.distCoeffs2 = None
###################################

# Optional
rigRect.computeRectificationMaps(alpha=0) # Alpha=0 may help removing unwanted border.

# Simply rectify two images (it takes care of distortion too)
img1_rect, img2_rect = rigRect.rectifyImages(img1, img2)

# Show images together
visImg = np.hstack((img1_rect, img2_rect))

visImg = cv2.resize(visImg, None, None, 0.1, 0.1) # Scaling for visualisation only...

# Draw some horizontal lines as reference
# (after rectification all horizontal lines are epipolar lines)
for y in [289, 332, 362]:
    cv2.line(visImg, (0,y), (visImg.shape[1],y), color=(0,0,255), thickness=2)

cv2.imshow('Rectified images', visImg)
cv2.waitKey(0)
cv2.destroyAllWindows()

print("Done!")

```

![nodist](https://github.com/decadenza/SimpleStereo/assets/30215028/aaf3ae06-5e3e-4015-bc62-f1076f2a40bd)

**Comment by felipe8066** [comment]
Hi @decadenza , I performed a new calibration taking into account the considerations you gave me, and I obtained better results. I used 207 images for calibration, resulting in a reprojection error of 1.16. The rectifications look much better than before. Now I'm facing the issue of how to use the information.

I'm working on a thesis project on a rectangular section laboratory channel, as shown in the following images:

![1_L](https://github.com/decadenza/SimpleStereo/assets/111525276/0f405a76-f341-42c6-93cd-7418aca95669)
![1_R](https://github.com/decadenza/SimpleStereo/assets/111525276/079e0ed3-e977-465c-aa14-1f030f71789f)

As you can see, the background is flat, and my idea is to determine the shape (heights) of the stones along the channel. However, the vertical wall is causing problems for data visualization. For this reason, I cropped the calibration images to keep only the inner part of the channel, and I obtained the following disparity map:

![Screenshot from 2024-01-25 21-11-09](https://github.com/decadenza/SimpleStereo/assets/111525276/f0ea4177-c255-4fe8-a662-eb4b0f13d538)

As you observe, the gravel is represented well, but I can't make the yellow flat part of the channel have consistent depths. Additionally, there seems to be a repetition or overlap between the left and right images, giving the impression that the calculated width is larger than the actual width. Have you encountered this problem? How can I visualize the "export.ply" file? I use MeshLab, but it only shows me a rectangle instead of a 3D representation of the rocks.

I would be eternally grateful if you could provide me with some guidance again to improve my results.

**Reply by decadenza** [reply]
What you see is mostly noise. The problem is not in the calibration which appears to be good now (a reprojection error of 1.16 at your high resolution is good).
The problem is in the stereo matching algorithm your are using. The ones implemented in SS are basic ones. You may want to play with their parameters or explore other solutions based on machine learning.

**Comment by felipe8066** [comment]
Thank you for your response; I'm glad to have a good fit in my calibration. Regarding your mention of using machine learning-based solutions, which ones could you recommend? I find it challenging to link the rig file obtained from calibration to the use of other methods. Upon investigation, I came across trained models like Hitnet and others used for depth determination. Is it possible to share any methods or solutions you know of that could be incorporated into my work if they prove to be effective?

Thank you again for your support.

**Reply by decadenza** [reply]
Hi there.
Nice work. I am glad that you managed to have the rig calibrated.
My research has been mostly on structured-light, so I didn't much work on passive stereo matching methods.

A good starting point could be looking here https://www.cvlibs.net/datasets/kitti/eval_scene_flow.php?benchmark=stereo
Bear in mind that those methods normally expect input images to be already rectified and/or other format restrictions.
So the working routine should be:

1) Capture calibration pictures -> Calibrate & rectify rig with SS -> Save stereo rig object

2) Capture your scene image(s) -> Undistort and rectify with SS -> Save new processed image(s)

3) Use processed image(s) with any stereo matching method of your choice. -> Obtain disparity map -> Save it.

4) Take the disparity map -> Geometrical (metric) reconstruction of your 3D scene -> Save to PLY (already in SS) or any other format (to be customised).

Now all the basic geometrical stuff is covered in SS, while the tasks in step 3 vary very differently. SS contains no advanced state-of-the-art methods for this. This is where you need to focus more to get good results.

Good luck.
